src/main/java/com/google/devtools/build/lib/cmdline/BazelModuleContext.java

```diff
@@ -44,6 +44,13 @@ public abstract class BazelModuleContext {
   /** The repository mapping applicable to the repo where the .bzl file is located in. */
   public abstract RepositoryMapping repoMapping();
 
+  /**
+   * The repository mapping applicable to the main repository, possibly without WORKSPACE repos or
+   * null. This is purely meant to support {@link Label#getDisplayFormForStarlark(StarlarkThread)}.
+   */
+  @Nullable
+  public abstract RepositoryMapping bestEffortMainRepoMapping();
```

Comment:
It's less work, but I also need to ensure that a top-level usage of `to_display_form()` returns an apparent repo name when the `.bzl` file is loaded (transitively) from a `BUILD` file, but a canonical name when loaded from from a module extension (as those don't track apparent names and I'm not even sure yet whether it should). Is that something that could be realized with a `StarlarkThread`-based approach?

Reply:
hmm. To track those usages, presumably you can just add relevant logic in `Label#getDisplayFormForStarlark` to use the `Label.RepoMappingRecorder` thread-local. But "it's less work" probably still stands.

Reply:
It's not just an omission, I think that there are valid reasons for *not* making apparent names available to module extensions: Build rules need apparent labels because they might have to pass them to actions as we see in Java and C++ rules, ultimately because users shouldn't use canonical labels in BUILD files. 

But module extensions don't really have a strong reason not to emit canonical labels (they work just fine with e.g. `bazel run` commands that update some files read by the extension) and the extra source of lockfile entry invalidation can be surprising to users.

I would prefer to take the cautious approach and release a Bazel version that supports this for BUILD files only. If we see that users actually ask for this to be available in module extensions, we can look into their actual use cases and reason about the tradeoff with more information. For example, I could see us making it so that `Label`s use apparent names in `print` and `fail` only, in which case we wouldn't need to track the mapping.

That's the long explanation for why I think `BazelModuleContext` is the right place at this time. In the future, when we do add support for extensions, I could move this logic over to the individual thread contexts.

Reply:
I think the argument is valid, but we may need to think about the naming. To me, `Label.to_display_form()` just looks like some method I should use in error messages, and it's strange that it does different things based on where I'm calling it from. I'd imagine it's also the same for a new(-ish) author just exploring the Label API; and it's also what the RELNOTES suggest.

But in reality, it looks like this is really for certain limited contexts where the rule author wants to show the user what to write in their BUILD files. I can't immediately come up with a succinct description for this scenario, but I wonder if you'd have better ideas.

Failing all that, I'm willing to accept this PR as is; but the various forms of labels are confusing enough already, and I'd like to reduce confusion if possible -- or at least think about it.

Reply:
I get your point and agree that we should take naming seriously. I'll think about this some more.

Reply:
Cc @shs96 @illicitonion Do you have a name in mind for this concept?

Reply:
`to_display_form` seems generally reasonable but I agree about it not being super clarifying... I guess this is really the "main workspace view of a label"? `relative_to_root_workspace` maybe?

Reply:
> To me, Label.to_display_form() just looks like some method I should use in error messages, and it's strange that it does different things based on where I'm calling it from.

I would say this is still true, even in contexts such as module extensions where we (currently? forever?) don't return apparent names: We still make main repo labels repo-relative, which is already a reason to prefer this for any kind of label "displayed" to a human.

If the way to get a canonical label literal wouldn't be `str(label)` but rather something like `label.canonical_form` and we could make the new method `label.display_form`, I would consider that pretty decent naming. As it stands, I also find it hard to come up with something better.

> `relative_to_root_workspace`

We would want to avoid the heavily overloaded term "workspace", but `relative_to_root_module` or `relative_to_main_repo` could work.